### ISSUE TITLE
restic prune: blob not found in repository

### ISSUE THREAD

**Issue by bzed** [restic/restic#3126]
Output of `restic version`
--------------------------
restic 0.11.0 (v0.11.0-149-g9a97095a) compiled with go1.15.5 on linux/amd64
(current master)

How did you run restic exactly?
-------------------------------
```
export GODEBUG=asyncpreemptoff=1
export RESTIC_REPOSITORY=/some/cifs/mount
export RESTIC_PASSWORD_FILE=somefile

restic prune
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

Windows file share / cifs mount.

Expected behavior
-----------------

restic should prune the repository, even when there are some files missing due to the repository being corrupt (for whatever reason), prune should finish. Or there should be a way to recover the repository so that woking/complete snapshots are left.

Actual behavior
---------------
```
1 bzed@conthink ~/.bin % restic prune                                                                               :(
repository a147f10d opened successfully, password is correct
loading all snapshots...
loading indexes...
finding data that is still in use for 12 snapshots
[0:00] 0.00%  0 / 12 snapshots
id 77310fe25bf7b6aba998c71f4c8f0f76916dd826c1a123abda1f3e0aff36b2b3 not found in repository
github.com/restic/restic/internal/repository.(*Repository).LoadBlob
	github.com/restic/restic/internal/repository/repository.go:158
github.com/restic/restic/internal/repository.(*Repository).LoadTree
	github.com/restic/restic/internal/repository/repository.go:790
github.com/restic/restic/internal/restic.FindUsedBlobs
	github.com/restic/restic/internal/restic/find.go:19
main.getUsedBlobs
	github.com/restic/restic/cmd/restic/cmd_prune.go:552
main.runPruneWithRepo
	github.com/restic/restic/cmd/restic/cmd_prune.go:174
main.runPrune
	github.com/restic/restic/cmd/restic/cmd_prune.go:151
main.glob..func19
	github.com/restic/restic/cmd/restic/cmd_prune.go:35
github.com/spf13/cobra.(*Command).execute
	github.com/spf13/cobra@v0.0.5/command.go:826
github.com/spf13/cobra.(*Command).ExecuteC
	github.com/spf13/cobra@v0.0.5/command.go:914
github.com/spf13/cobra.(*Command).Execute
	github.com/spf13/cobra@v0.0.5/command.go:864
main.main
	github.com/restic/restic/cmd/restic/main.go:98
runtime.main
	runtime/proc.go:204
runtime.goexit
	runtime/asm_amd64.s:1374
```

Steps to reproduce the behavior
-------------------------------

Delete some random files, I'd guess.

Do you have any idea what may have caused this?
-----------------------------------------------

Broken cifs server, broken internet connection, no idea really. But there should be a way to recover the data.

Do you have an idea how to solve the issue?
-------------------------------------------
Not really.
Unfortunately I have to remove the repo, because its just growing with every backup and I can't prunr it anymore.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
I have about 10 TB of backup managed by restic, no issues so far - this is the only one and probably just happening due to using a cifs mount trough vpm from homeoffice with a maybe buggy windows server. Never had any other issue so far! :+1:

**Comment by aawsome** [comment]
The problem is that some tree blobs are missing, so `prune` cannot determine what is still needed and what not.
If you use an unpatched restic, you have to find out the affected snapshots and remove them. Then prune works again.

If you are fine with using a patched version, you can try out #2876

**Comment by bzed** [comment]
I've removed all but the snapshtots from the few last days, did not expect to have an error there as the backup was happy.
Unfortunately I can't test it anymore as my co-workers prodded me to clean up the backupspace.

**Comment by MichaelEischer** [comment]
@bzed Were these twelve snapshots created using restic 0.11? Did any of these backup runs fail (in case you have logs)? For your prune error to show up either a pack file from the data folder must have been lost or an index file (`rebuild-index` can help in these cases). However either cause should normally cause a backup operation to fail. How often do you run backup and prune?

**Comment by bzed** [comment]
> @bzed Were these twelve snapshots created using restic 0.11?

For sure.

>Did any of these backup runs fail (in case you have logs)?

Definitely. Don't have logs, but this is the backup of my laptop. It is put to sleep at random times, rebootet without taking care of running jobs,..... and so on.

> For your prune error to show up either a pack file from the data folder must have been lost or an index file (`rebuild-index` can help in these cases).

rebuild-index did not help.

> However either cause should normally cause a backup operation to fail. How often do you run backup and prune?

Backup: usually daily. Forget and prune once a week.

**Comment by MichaelEischer** [comment]
@bzed The interruptions at random points shouldn't be a problem. At least the `backup` command shouldn't have any problems coping with that. I'm using restic with sftp or rclone over ssh on my laptop (I also don't check whether a backup is running prior to interrupting the network connection or similar) and haven't had any problems with repository corruption. So I'm going to speculate a bit about what could be different with your setup.

I'm currently triggering `prune` runs manually. That usually requires calling `unlock` first, as the interrupted backups leave stale locks behind.

When automatically calling `unlock` without further synchronization with the backup there could be a weird race condition. Putting your laptop to sleep could allow a lock created by the backup run to expire while the laptop is asleep, which could then be deleted by `unlock` after waking up. If the CIFS mount automatically reconnects after waking up the backup would still progress further, while (now that the lock is gone) a prune run can also start in parallel to the backup, which with bad timing could remove the top-level tree of a snapshot. If that scenario occurs, then the backup run would complain about being unable to delete a lock file. (That scenario would be fixed by #2715)

Other options would be that CIFS reports a file write/sync/close as successful while it isn't. Or that the CIFS mount disappears temporarily, restic stores a few data files there which are later hidden when the mountpoint is activated again.

**Comment by MichaelEischer** [comment]
This issues is too old to get anything usable out of it. The lock file handling as well as prune have been improved in the meantime.
